firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/FirestoreMapEncoder.kt

```diff
@@ -0,0 +1,210 @@
+ * [FirestoreListEncoder].
+ *
+ * @param map the nested map that records the encoded result of the original @Serializable object.
+ * @param depth the current encoding depth inside of the nested map.
+ * @param encodedList the encoded result of the current nested list.
+ */
+class FirestoreListEncoder(
+    private val map: MutableMap<Int, MutableMap<String, Any?>> = mutableMapOf(),
+    private var depth: Int = 0,
+    private val encodedList: MutableList<Any?> = mutableListOf()
+) : AbstractEncoder() {
+
+    override val serializersModule: SerializersModule = EmptySerializersModule
+
+    private var elementIndex: Int = 0
+
+    override fun encodeValue(value: Any) {
+        encodedList.add(value)
+        elementIndex++
+    }
+    override fun encodeNull() {
+        encodedList.add(null)
+        elementIndex++
+    }
+
+    /**
+     * Recursively build the nested map when the element inside of the nested list has its own
+     * structure (i.e. a list of nested custom @Serializable object)
+     *
+     * <p>Note: Firestore does not support read/write high dimensional nested lists; therefore,
+     * list's content must be either primitive type elements or @Serializable objects.
+     * IllegalArgumentException will be thrown if list contains elements of other types.
+     *
+     * @param descriptor the [SerialDescriptor] of the @Serializable object at current depth.
+     * @return a [FirestoreMapEncoder].
+     */
+    override fun beginStructure(descriptor: SerialDescriptor): CompositeEncoder {
+        when (descriptor.kind) {
+            StructureKind.CLASS -> {
+                val nextDepth = depth + 1
+                map[nextDepth] = mutableMapOf()
+                encodedList.add(map[nextDepth])
+                return FirestoreMapEncoder(map, nextDepth, descriptor.elementNames.toMutableList())
+            }
+            else -> {
+                throw IllegalArgumentException(
+                    "Incorrect format of nested object provided: <$descriptor.kind>"
+                )
+            }
+        }
+    }
+}
+
+/**
+ * Returns the encoded contents of a @Serializable object converted to a nested map.
+ *
+ * @param serializer The SerializationStrategy of the @Serializable Kotlin object.
+ * @param value The @Serializable Kotlin object.
+ * @return The encoded nested map of a @Serializable Kotlin object.
+ */
+fun <T> encodeToMap(serializer: SerializationStrategy<T>, value: T): MutableMap<String, Any?> {
+    val encoder = FirestoreMapEncoder()
+    encoder.encodeSerializableValue(serializer, value)
+    return encoder.serializedResult()
+}
+
+/**
+ * Returns the encoded contents of a @Serializable object converted to a nested map.
+ *
+ * @param value The @Serializable Kotlin object.
+ * @return The encoded nested map of a @Serializable Kotlin object.
+ */
+inline fun <reified T> encodeToMap(value: T): MutableMap<String, Any?> =
+    encodeToMap(serializer(), value)
```

Comment:
for my own understanding: where is the `serializer()` coming from?

Reply:
> `serializer(typeOf<T>())`

This could be an issue as it looks like it uses reflection and finds the serializer of the type by looking up the `serializer()` method by name on the class. This means that this code won't work in the face of proguard/obfuscation unless the developer adds keeps on their serializable classes(which is not something kotlinx.serialization requires). Something to think about

Reply:
Friendly ping @vkryachko

Reply:
good to know, I was under the impression that kotlinx did not require any proguard configs and worked when obfuscated